### DISCUSSION TITLE
Customize login controller with api

### DISCUSSION THREAD

**Discussion by jlopes90** [codeigniter4/shield#233] [Q&A] [answered]
I wanted to login controller with api, but I've been testing `loggenIn` on the "dashboard" site. Says it's false, I don't know why. Without api it works `loggenIn` is true.

### for example: current login controller without api
```php
$credentials = [
    'email'    => $this->request->getPost('email'),
    'password' => $this->request->getPost('password')
];

$loginAttempt = auth()->attempt($credentials);

if (! $loginAttempt->isOK()) {
    return redirect()->back()->with('error', $loginAttempt->reason());
}

return redirect()->to('dashboard');
```
dashboard: `loggenIn` is true.

### for example: customize login controller with api
```php
$credentials = [
    'email'    => $this->request->getPost('email'),
    'password' => $this->request->getPost('password')
];

$loginAttempt = auth()->attempt($credentials);

if (! $loginAttempt->isOK()) {
    return return $this->respond([
        'success' => false
    ]);
}

return return $this->respond([
    'success' => true,
    'url' => 'dashboard'
]);
```
dashboard: `loggenIn` is false.

I have no idea. Why `redirect()` says true and `$this->respond()` says false in `loggenIn`?

**Comment by kenjis** [comment]
How did you check the `loggedIn()` on the "dashboard" ?

**Reply by jlopes90** [reply]
> How did you check the `loggedIn()` on the "dashboard" ?

```
class Dashboard extends BaseController
{
    public function index()
    {
        dd(auth()->loggedIn());

        if (!auth()->loggedIn()) {
            return redirect()->to(route_to('login'));
        }
        
        ...
    }
}
```

**Reply by kenjis** [reply]
I cannot reproduce.

```diff
--- a/app/Config/Routes.php
+++ b/app/Config/Routes.php
@@ -37,6 +37,9 @@ $routes->set404Override();
 // route since we don't have to scan directories.
 $routes->get('/', 'Home::index');
 
+$routes->get('dashboard', 'Dashboard::index');
+$routes->post('login', 'Login::index');
+
 service('auth')->routes($routes);
 
 /*
```
```php
<?php

namespace App\Controllers;

use CodeIgniter\API\ResponseTrait;

class Login extends BaseController
{
    use ResponseTrait;

    public function index()
    {
        $credentials = [
            'email' => $this->request->getPost('email'),
            'password' => $this->request->getPost('password')
        ];

        $loginAttempt = auth()->attempt($credentials);

        if (!$loginAttempt->isOK()) {
            return $this->respond([
                                      'success' => false
                                  ]);
        }

        return $this->respond([
                                  'success' => true,
                                  'url' => 'dashboard'
                              ]);
    }
}
```
```php
<?php

namespace App\Controllers;

class Dashboard extends BaseController
{
    public function index()
    {
        d(auth()->loggedIn());
    }
}
```

http://localhost:8080/login
![Screenshot 2022-06-12 14 23 16](https://user-images.githubusercontent.com/87955/173217050-dddbe1ec-2c28-4e82-997f-8f6e19da324b.png)

http://localhost:8080/dashboard
![Screenshot 2022-06-12 14 23 41](https://user-images.githubusercontent.com/87955/173217051-f1cfff66-6edc-480b-ae5d-cbb6b28c98c8.png)

**Reply by jlopes90** [reply]
I already did the same as you did, but the result is false.

Don't use `remember`, because the result is always true.

Login:
https://user-images.githubusercontent.com/50026015/173229973-73d9061f-dcc3-4cbf-86e0-a62fde94b419.png

Dashboard:
https://user-images.githubusercontent.com/50026015/173229815-75dca407-21d8-45be-af45-c5331c6814e0.png

**Reply by jlopes90** [reply]
wait a sec

**Reply by jlopes90** [reply]
The error continues...

I've been testing something else, it's deleting the "cookies".

1 - open "login" and delete all cookies and enter an account.
2 - opening "dashboard" says `loggenIn` is true.
3 - open "logout", then open "dashboard" to see result `loggenIn` says is true.

1 - open "login" and delete all cookies and refresh login.
2 - updated "login" and not delete all cookies and enter an account.
3 - opening "dashboard" says `loggenIn` is false.

If you delete all cookies, `loggenIn` is always false.

**Reply by kenjis** [reply]
Logout does not work now because of a bug. See #225

**Reply by jlopes90** [reply]
> Logout does not work now because of a bug. See #225

Oops! I had already modified logout, but then I did composer update today, that's why it replaces itself.

I'll see now.

**Reply by jlopes90** [reply]
> 1 - open "login" and delete all cookies and enter an account.
> 2 - open "dashboard" says `loggenIn` is true.
> 3 - open "logout", then open "dashboard" to see result `loggenIn` says is ~~true~~ false.
> 
> 1 - open "login" and delete all cookies and refresh login
> 2 - updated "login" and not delete all cookies and enter an account.
> 3 - open "dashboard" says `loggenIn` is false.

**Reply by jlopes90** [reply]
If you don't delete all cookies, login an account and loggenIn is always false.

**Reply by kenjis** [reply]
I can't reproduce. The same result as before.

1 - open "login" page and delete all cookies and refresh login page.
2 - updated "login" page and not delete all cookies and enter email/password.
3 - open "dashboard" says loggedIn is true.

**Reply by jlopes90** [reply]
Strange, mine is different. I can't solve.

I tested the "postman" program and it works but in browsers it doesn't work.

**Reply by jlopes90** [reply]
The problem is pending. I will use without api.

**Comment by datamweb** [comment]
@jlopes90  **return return** !!

**Comment by jlopes90** [comment, marked as answer]
I just updated  `codeigniter 4.2.3` and the  `shield develop` too and tested it with api and it's already working.
